Added ConfigJson class.

sethraymond opened the pull request:
ConfigJson acts as a wrapper for the Speedtest options. Easy method
for toggling pingOnly test option.

Commit:
Added ConfigJson class.

ConfigJson acts as a wrapper for the Speedtest options. Easy method
for toggling pingOnly test option.

matrumz commented on blameTheISP.py:
```diff
@@ -4,47 +4,87 @@
+        with open(self.path, 'r') as config_json:
+            self.data = json.load(config_json)
+
+    def toggle_ping_only():
```
I see what this function does, but it doesn't look like it's used at all. What are you intending this to be used for in the future?

sethraymond replied:
For later use - not necessarily written for "right now," it's written to be used if necessary in the future. It's best practice that if you have a dump/write method, you should include a load/read method.

matrumz replied:
Eh? This question was on the "toggle_ping_only" method.

And I think you missed some questions :stuck_out_tongue_winking_eye:

sethraymond replied:
I see it as asking on the `load` method.

Again, it's more for later use. With whatever future API you're using, when you want to toggle the `pingOnly` option, you simply load the JSON, toggle, and then write it back. It's cleaner than writing out `ConfigJson.data['pingOnly'] = not ConfigJson.data['pingOnly']`, especially if `ConfigJson` lives in its own .py file.

matrumz replied:
Okay. I think I'll scratch this when I pull it in, sorry. The API will be a JS web-service controller that writes directly to the JSON file. Details on the API to come.

matrumz commented on blameTheISP.py:
```diff
@@ -4,47 +4,87 @@
+
+    #speed_test_opts = ConfigJson('~/Blame-The-ISP/blameTheISP.config.json')
+
+    with open(os.path.join(os.path.dirname(os.path.realpath(__file__)), "blameTheISP.config.json"), "r") as config_file: config_data = json.load(config_file)
```
Why no new-line?

sethraymond replied:
This was actually as it was when I got the code. The only reason it "changed" is because I wrapped it in the `main()` function.

matrumz replied:
Not quite: see image below. I just didn't know if this was a particular style choice... a newline must've gotten lost during conversion. No biggie... unless the whitespace-sensitivity of Python is going to complain.

![image](https://user-images.githubusercontent.com/9788609/31352084-692a7b5a-acfb-11e7-937f-55eb73bb754c.png)

sethraymond replied:
That's odd - definitely wasn't there when I opened it in vim. It's valid syntax to have it all on one line, though it breaks PEP8 convention and running the `pep8` checker will yell at you for it.

matrumz replied:
Well I DID develop this in Windows... VSCode was supposed to take care of line-endings, though.

matrumz commented on blameTheISP.py:
```diff
@@ -4,47 +4,87 @@
+
+def main():
+
+    #speed_test_opts = ConfigJson('~/Blame-The-ISP/blameTheISP.config.json')
```
You went through the trouble of creating the class... why not use it?

sethraymond replied:
1.) The class isn't actually instantiated here, it's commented out.
2.) This commit isn't necessarily the final version, it's just a framework. If this line wasn't commented out, you would access variables via `speed_test_opts.data[key]`, which is somewhat ugly. Maybe writing `__getitem__` and `__setitem__` methods or `ConfigJson` would be a cleaner way of doing it?

matrumz replied:
I like the `__get/setitem__` idea... do you want to submit another PR that will actually count towards the hackathon?

sethraymond replied:
Can do. the `__getitem__` method will essentially translate into `val = ConfigJson[key]` and `__setitem__` to `ConfigJson[key] = val`

sethraymond replied:
Eliminates the whole `.data` business.

matrumz replied:
:thumbsup: I'll cancel this PR and let you make your mods for re-submission while I work on the API.

--EDIT--
Nevermind (I'm still learning about PRs). You can just create another commit on this PR branch and I'll be pulling in both.

matrumz reviewed (commented):
General questions.

Commit:
Added __getitem__ and __setitem__ methods.

Main function is now using the ConfigJson class. Next step is to
probably create a Speedtest class that encapsulates the Speedtest
class and the options file, maybe also the database.

matrumz commented on blameTheISP.py:
```diff
@@ -41,32 +47,30 @@ def toggle_ping_only():
-    #speed_test_opts = ConfigJson('~/Blame-The-ISP/blameTheISP.config.json')
-
-    with open(os.path.join(os.path.dirname(os.path.realpath(__file__)), "blameTheISP.config.json"), "r") as config_file: config_data = json.load(config_file)
+    speed_test_opts = ConfigJson('~/Blame-The-ISP/blameTheISP.config.json')
```
This assumes the user's directory structure. The config file (by default) should be found as a sibling to this py file, regardless of where that is.

sethraymond replied:
Does the sqlite3 database live in that same directory?

matrumz replied:
No, the database should be anywhere, as it is likely to be consumed by other code.

matrumz reviewed (commented):
I like where this is going... though I do have one more recommended edit to your PR. See comments.

Commit:
Should check script path.

Please test to see if this is cron safe.

matrumz commented on blameTheISP.py:
```diff
@@ -47,7 +48,11 @@ def toggle_ping_only():
+    # Probably a better way to do this, but I think this is cron safe.
+    speedtest_dir = os.path.realpath(sys.argv[0])
+    opts_path = '/'.join(speedtest_dir.split('/')[:-1])
+    opts_path += '/blameTheISP.config.json'
```
Might I suggest replacing lines 52-54 with:
`opts_path = os.path.join(os.path.abspath(os.path.dirname(__file__)), 'blameTheISP.config.json')`
OR
`speedtest_dir = os.path.abspath(os.path.dirname(__file__))`
`opts_path = os.path.join(speedtest_dir, 'blameTheISP.config.json'`

These methods do not assume OS path notations ('/' vs '\'), don't require the import of sys, and are proven to be cron-safe.

matrumz reviewed (changes requested):
Assumes Linux paths. More concise and more hands-off methods available.

matrumz commented:
@sethraymond  Thanks for the nice code! You're :sparkles:awesome:sparkles:!
